nucypher/blockchain/eth/sol/source/contracts/PREApplication.sol

```diff
@@ -0,0 +1,652 @@
+        }
+
+        if (info.authorized == 0) {
+            _stakingProviderFromOperator[info.operator] = address(0);
+            info.operator = address(0);
+            info.operatorConfirmed == false;
+        }
+    }
+
+    //-------------------------Main-------------------------
+    /**
+    * @notice Returns staking provider for specified operator
+    */
+    function stakingProviderFromOperator(address _operator) public view override returns (address) {
+        return _stakingProviderFromOperator[_operator];
+    }
+
+    /**
+    * @notice Returns operator for specified staking provider
+    */
+    function getOperatorFromStakingProvider(address _stakingProvider) public view returns (address) {
+        return stakingProviderInfo[_stakingProvider].operator;
+    }
+
+    /**
+    * @notice Get all tokens delegated to the staking provider
+    */
+    function authorizedStake(address _stakingProvider) public view override returns (uint96) {
+        return stakingProviderInfo[_stakingProvider].authorized;
+    }
+
+    /**
+    * @notice Get the value of authorized tokens for active providers as well as providers and their authorized tokens
+    * @param _startIndex Start index for looking in providers array
+    * @param _maxStakingProviders Max providers for looking, if set 0 then all will be used
+    * @return allAuthorizedTokens Sum of authorized tokens for active providers
+    * @return activeStakingProviders Array of providers and their authorized tokens.
+    * Providers addresses stored as uint256
+    * @dev Note that activeStakingProviders[0] is an array of uint256, but you want addresses.
+    * Careful when used directly!
+    */
+    function getActiveStakingProviders(uint256 _startIndex, uint256 _maxStakingProviders)
+        external view returns (uint256 allAuthorizedTokens, uint256[2][] memory activeStakingProviders)
+    {
+        uint256 endIndex = stakingProviders.length;
+        require(_startIndex < endIndex, "Wrong start index");
+        if (_maxStakingProviders != 0 && _startIndex + _maxStakingProviders < endIndex) {
+            endIndex = _startIndex + _maxStakingProviders;
+        }
+        activeStakingProviders = new uint256[2][](endIndex - _startIndex);
+        allAuthorizedTokens = 0;
+
+        uint256 resultIndex = 0;
+        for (uint256 i = _startIndex; i < endIndex; i++) {
+            address stakingProvider = stakingProviders[i];
+            StakingProviderInfo storage info = stakingProviderInfo[stakingProvider];
+            uint256 eligibleAmount = info.authorized - info.deauthorizing;
+            if (eligibleAmount < minAuthorization || !info.operatorConfirmed) {
+                continue;
+            }
+            activeStakingProviders[resultIndex][0] = uint256(uint160(stakingProvider));
+            activeStakingProviders[resultIndex++][1] = eligibleAmount;
+            allAuthorizedTokens += eligibleAmount;
+        }
+        assembly {
```

Comment:
also not sure what this is doing, I know solidity has the ability to run assembly, but I have no idea when/why you might want to do that

Reply:
There's no dynamic memory arrays. This line is used to change/decrease array length for fixed memory array. Otherwise this array will have a lot of empty elements which will slow down a call. I know that using assembly is not preferable but for my taste it fits here perfectly

Reply:
ok cool, makes sense!